### PR TITLE
feat: build pipeline visualizer

### PR DESCRIPTION
**Pull request by plyr4**
requires server changes from https://github.com/go-vela/server/pull/995

closes https://github.com/go-vela/community/issues/71

- adds elm DOT module
- adds [graphviz](https://graphviz.org/doc/info/lang.html) and [d3](https://d3js.org/) via npm
- adds new page `org/repo/1/graph` for build tab titled `Visualize`
- adds outbound interop for rendering the actual graph
- adds inbound interop for handling graph interactions

the code added will retrieve a graph object from the server, generate a DOT string in Elm, generate the html using Graphviz, then bootstrap style and interactivity using [d3](https://d3js.org/) and scss.

inspiration was taken from the [d3-graphviz](https://github.com/magjac/d3-graphviz) by https://github.com/magjac and various CI tools.

## Demo

https://github.com/go-vela/ui/assets/48764154/e72136e5-7b22-4393-b88b-7f4b4717c46d

#### Light Theme

needs some work, but so does the rest of light theme.

Screenshot 2023-10-27 at 9 46 06 AM

### CODE REVIEW

**Comment by plyr4** [comment]
> Have you considered generating the DOT graph representation server-side? Would doing that prevent the links from the Graph to the relevant step/stage/service from working?

great question! I did some research that I detailed in [the issue](https://github.com/go-vela/community/issues/71#issuecomment-1003212679) but short-story is the DOT representation needs to be aware of some html/css prior to generating the layout using graphviz (see the [`nodeLabel`](https://github.com/go-vela/ui/blob/feat/viz-d3/src/elm/Pages/Build/Graph/DOT.elm#L155-L160) function). also it felt sort of wrong to be hardcoding html/css in the server code, especially if i need to duplicate those classes in the UI to style them anyways.

plus, to make it worse, graphviz is a c library so the ideal solution for Go requires enabling `CGO` :( which i would like to avoid.

**Comment by cognifloyd** [comment]
> > Have you considered generating the DOT graph representation server-side? Would doing that prevent the links from the Graph to the relevant step/stage/service from working?
> 
> great question! I did some research that I detailed in [the issue](https://github.com/go-vela/community/issues/71#issuecomment-1003212679) but short-story is the DOT representation needs to be aware of some html/css prior to generating the layout using graphviz (see the [`nodeLabel`](https://github.com/go-vela/ui/blob/feat/viz-d3/src/elm/Pages/Build/Graph/DOT.elm#L155-L160) function). also it felt sort of wrong to be hardcoding html/css in the server code, especially if i need to duplicate those classes in the UI to style them anyways.
> 
> plus, to make it worse, graphviz is a c library so the ideal solution for Go requires enabling `CGO` :( which i would like to avoid (see [issue comment](https://github.com/go-vela/community/issues/71#issuecomment-1006065219)).

Ah. I missed that discussion - oops. Sounds like client-side DOT is a great tool here. I like where this is going :smile:.

**Review by wass3r — COMMENTED** [review]
super cool. was playing with it a bit before looking through code. i noticed that when a step errors (not a failure), the graph doesn't correctly reflect it. the stage the error'd step occurred in is rendered with green outline and the step itself doesn't render a status:

image

should be able to reproduce with something like (will error due to duplicate step name in the stage):

```yaml
version: "1"

stages:
  this:
    steps:
      - name: sleeping
        image: alpine:latest
        commands:
          - sleep 5
  thing:
    needs: this
    steps:
      - name: sleeping
        image: alpine:latest
        commands:
          - sleep 20
      - name: sleeping
        image: alpine:latest
        commands:
          - sleep 4
```

aside from that, seems to be working well so far 🥳

**Comment by wass3r** [comment]
see the fix, thank you! only feedback i have is that i sort of expected the whole container to get marked as failed as it happens in your video from the description:
image

in my case it stayed green even though one of the steps error'd and caused the build to fail. i think the only scenario i can see where the whole container should still be green is if they had ruleset: continue: true in there?

**Comment by plyr4** [comment]
sure I'll check out that scenario. that logic is controlled by the BuildGraph endpoint so the best we can do is tweak the OverallStageStatus function in the server code to be better at determining "stage status" and then the UI will just reflect whatever it returns

File: src/elm/Pages/Build/Graph/DOT.elm [hunk 1]
```diff
@@ -0,0 +1,505 @@
...
+            node.name
+
+        steps =
+            List.sortBy .id node.steps
+
+        table content =
+            "<table "
+                ++ escapeAttributes nodeLabelTableAttributes
+                ++ ">"
+                ++ String.join "" content
```
Comment (wass3r) [inline src/elm/Pages/Build/Graph/DOT.elm:181]:
```suggestion
                ++ String.concat content
```

File: src/elm/Pages/Build/Graph/DOT.elm [hunk 2]
```diff
@@ -0,0 +1,505 @@
...
+nodeLabel model repo build node showSteps =
+    let
+        label =
+            node.name
+
+        steps =
+            List.sortBy .id node.steps
+
+        table content =
+            "<table "
```
Comment (wass3r) [inline src/elm/Pages/Build/Graph/DOT.elm:178]: probably not much we can do here, but too bad we have to string concatenate html 😢

**Review by wass3r — DISMISSED** [review]
one generic note: make sure to add license headers and fix outdated ones. otherwise, some inline comments, suggestions, etc.

File: src/scss/_graph.scss [hunk 3]
```diff
@@ -0,0 +1,475 @@
...
+}
+
+.d3-build-graph-edge-tip {
+  fill: var(--color-gray);
+  stroke: var(--color-gray);
+}
+
+// end: classes controlled by d3
+
+// start: theme overrides
```
[the remark this answers is not in the document]
Comment (plyr4) [reply src/scss/_graph.scss:406]: yeah agreed, i'll move it

File: src/scss/_graph.scss [hunk 4]
```diff
@@ -0,0 +1,475 @@
...
+
+  fill: var(--color-bg-dark);
+  stroke: var(--color-gray);
+
+  .-pending {
+    stroke-dasharray: 7, 4;
+  }
+
+  .-running {
+    stroke: var(--color-yellow);
```
[the remark this answers is not in the document]
Comment (plyr4) [reply src/scss/_graph.scss:109]: agreed, i'll do another pass and collapse shared style

File: src/elm/Pages/Build/Graph/View.elm [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+                    , text "- selected"
+                    ]
+                , li []
+                    [ buildVizLegendEdge [ Svg.Attributes.class "-pending" ]
+                    , text "- pending"
+                    ]
+                , li []
+                    [ buildVizLegendEdge [ Svg.Attributes.class "-finished" ]
+                    , text "- complete"
+                    ]
```
[the remark this answers is not in the document]
Comment (plyr4) [reply src/elm/Pages/Build/Graph/View.elm]: true! let me add some margin and remove the `-`

File: src/elm/Pages/Build/Graph/View.elm [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+                ]
+            , case model.repo.build.graph.graph of
+                RemoteData.Success _ ->
+                    -- dont render anything when the build graph draw command has been dispatched
+                    text ""
+
+                RemoteData.Failure _ ->
+                    -- todo: make this cuter
+                    text "Error loading build graph... Please try again"
```
[the remark this answers is not in the document]
Comment (plyr4) [reply src/elm/Pages/Build/Graph/View.elm:183]: im having trouble defining "cuter" 😅

File: src/elm/Pages/Build/Graph/Interop.elm [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+import Routes exposing (Route(..))
+import Vela exposing (encodeBuildGraphRenderData)
+
+
+{-| renderBuildGraph : takes partial build model and render options, and returns a cmd for dispatching a graphviz+d3 render command
+-}
+renderBuildGraph : BuildModel.PartialModel a -> Bool -> Cmd msg
+renderBuildGraph model centerOnDraw =
+    -- rendering the full graph requires repo, build and graph
+    case model.repo.repo of
```
[the remark this answers is not in the document]
Comment (plyr4) [reply src/elm/Pages/Build/Graph/Interop.elm:16]: good call

File: src/elm/Pages/Build/Graph/View.elm (hunk 6 above)
[the remark this answers is not in the document]
Comment (plyr4) [reply src/elm/Pages/Build/Graph/View.elm:183]: actually i did get some feedback on this earlier to re-use some toasty style and create a banner. i can do another pass at this

File: src/elm/Pages/Build/Graph/DOT.elm (hunk 2 above)
Comment (plyr4) [reply src/elm/Pages/Build/Graph/DOT.elm:178]: i mean i 100% agree the code screams "hacky" 

im open to alternatives, but its not _really_ html, its "html-like" strings, following the graphviz docs:
https://graphviz.org/docs/attrs/label/

Comment (wass3r) [reply src/elm/Pages/Build/Graph/DOT.elm:178]: yup, no worries. necessary evil

File: src/static/index.d.ts [hunk 8]
```diff
@@ -95,3 +97,6 @@ export type ToElm<T> = {
  *
  */
 export type Theme = 'theme-light' | 'theme-dark';
+
+export type GraphData = {};
+export type GraphInteraction = {};
```
[the remark this answers is not in the document]
Comment (plyr4) [reply src/static/index.d.ts]: these should be accurate now, i also made everything camelcase, had it snake case to make it a bit more obvious when i was debugging. thanks for pointing it out

File: src/elm/Pages/Build/Graph/DOT.elm [hunk 9]
```diff
@@ -0,0 +1,510 @@
...
+-}
+baseGraphStyles : Rankdir -> Styles
+baseGraphStyles rankdir =
+    { rankdir = rankdir
+    , graph =
+        escapeAttributes
+            [ ( "bgcolor", DefaultEscape "transparent" )
+            , ( "splines", DefaultEscape "ortho" )
+
+            -- , ( "compound", DefaultEscape "true" )
```
Comment (wass3rw3rk) [inline src/elm/Pages/Build/Graph/DOT.elm:311]: do we need to keep this?

File: src/static/index.d.ts [hunk 10]
```diff
@@ -95,3 +95,33 @@ export type ToElm<T> = {
...
+};
+
+export type GraphInteraction = {
+  /** @property eventType: string */
+  eventType: string;
+  /** @property href: string */
+  href: string;
+  /** @property nodeID: string */
+  nodeID: string;
+};
```
Comment (wass3rw3rk) [inline src/static/index.d.ts:127]: :heart:

**Review by wass3rw3rk — APPROVED** [review]

**Review by wass3rw3rk — DISMISSED** [review]

File: src/elm/Pages/Build/Graph/DOT.elm (hunk 9 above)
Comment (plyr4) [reply src/elm/Pages/Build/Graph/DOT.elm:311]: oh, just saw this, no, i was just playing around with compound graphs. let me remove that before it ends up being there forever

**Review by wass3rw3rk — APPROVED** [review]

**Review by ecrupper — APPROVED** [review]

**Pull request merged**